crypto/ml_kem/ml_kem.c

```diff
@@ -0,0 +1,1567 @@
+        scalar_ntt(a++);
+}
+
+/* In-place inverse NTT transform of a vector */
+static void vector_inverse_ntt(scalar *a, int rank)
+{
+    while (rank-- > 0)
+        scalar_inverse_ntt(a++);
+}
+
+static void
+matrix_mult(scalar *out, const scalar *m, const scalar *a, int rank)
+{
+    const scalar *ar;
+    int i, j;
+
+    for (i = rank; i-- > 0; ++out) {
+        scalar_mult(out, m++, ar = a);
+        for (j = rank - 1; j > 0; --j)
+            scalar_mult_add(out, m++, ++ar);
+    }
+}
+
+static void
+matrix_mult_transpose(scalar *out, const scalar *m, const scalar *a, int rank)
+{
+    const scalar *mc = m, *mr, *ar;
+    int i, j;
+
+    for (i = rank; i-- > 0; ++out)  {
+        scalar_mult(out, mr = mc++, ar = a);
+        for (j = rank; --j > 0; )
+            scalar_mult_add(out, (mr += rank), ++ar);
+    }
+}
+
+/*
+ * Generates a secret vector by using |cbd| with the given seed to generate
+ * scalar elements and incrementing |counter| for each slot of the vector.
+ */
+static __owur
+int gencbd_vector(scalar *out, cbd_t cbd, uint8_t *counter,
+                  const uint8_t seed[ML_KEM_RANDOM_BYTES], int rank, mctx *ctx)
+{
+    uint8_t input[ML_KEM_RANDOM_BYTES + 1];
+
+    memcpy(input, seed, ML_KEM_RANDOM_BYTES);
+    while (rank-- > 0) {
+        input[ML_KEM_RANDOM_BYTES] = (*counter)++;
+        if (!cbd(out++, input, ctx))
+            return 0;
+    }
+    return 1;
+}
+
+/*-
+ * Expands the matrix from a seed for key generation and for encaps-CPA.
+ * NOTE: FIPS 203 matrix "A" is the transpose of this matrix, computed
+ * by appending the (i,j) indices to the seed in the opposite order!
+ *
+ * Where FIPS 203 computs $t = A * s + e$, we use the transpose of "m".
+ */
+static __owur
+int matrix_expand(scalar *out, const uint8_t rho[ML_KEM_RANDOM_BYTES],
+                  int rank, mctx *ctx)
+{
+    uint8_t input[ML_KEM_RANDOM_BYTES + 2];
+    int i, j, ret = 0;
+    EVP_MD_CTX *mdctx = EVP_MD_CTX_new();
+
+    if (mdctx == NULL)
+        goto end;
+    memcpy(input, rho, ML_KEM_RANDOM_BYTES);
+    for (i = 0; i < rank; i++) {
+        for (j = 0; j < rank; j++) {
+            input[ML_KEM_RANDOM_BYTES] = i;
+            input[ML_KEM_RANDOM_BYTES + 1] = j;
+            if (!EVP_DigestInit_ex(mdctx, ctx->shake128_cache, NULL)
+                || !EVP_DigestUpdate(mdctx, input, sizeof(input))
+                || !sample_scalar(out++, mdctx))
+                goto end;
+        }
+    }
+
+    ret = 1;
+end:
+    EVP_MD_CTX_free(mdctx);
+    return ret;
+}
+
+/*
+ * FIPS 203, Section 5.2, Algorithm 14: K-PKE.Encrypt.
+ *
+ * Encrypts a message with given randomness to the ciphertext in |out|. Without
+ * applying the Fujisaki-Okamoto transform this would not result in a CCA
+ * secure scheme, since lattice schemes are vulnerable to decryption failure
+ * oracles.
+ *
+ * Caller passes storage for |y|, |e1| and |u|.
+ */
+static __owur
+int encrypt_cpa(uint8_t *out, const uint8_t *message, const scalar *m,
```

Comment:
Maybe some comments related to your extra input params could be given.. m = A` and t are stored rather than calculated as part of this alg.

Reply:
Please state the assumptions about in/out sizes (especially out).

Reply:
Done wherever the sizes are actually constant.